rfcs/20190430-tokenization-conventions.md

```diff
@@ -0,0 +1,352 @@
+
+#### 1. String inputs - Speed vs Simplicity {#1-string-inputs-speed-vs-simplicity}
+
+The input to each op is a Tensor (or RaggedTensor). For the common case, the
+type of these tensors will be UTF-8 strings, but implementers may choose for
+their tokenizers to also work on integers. By accepting strings, we wish to make
+adoption and usage as easy as possible, even if comes at a slight performance
+hit by working only with Unicode codepoints.
+
+#### 2. Ragged outputs {#2-ragged-outputs}
+
+The number of tokens created from tokenizing a string is unknown. For this
+reason, it is impossible to fully tokenize and output a normal tensor with a
+uniform shape for a batch of varying strings. Thus, it is expected that each
+output will be ragged.
+
+To allow the caller to know which groups of tokens belong to each string, the
+innermost ragged dimension will be tokens for the originating string. This means
+that the shape of the output will have an additional dimension when compared to
+the input. Example:
+
+```python
+>>> whitespace_tokenize(["This is great!", "Awesome!"])
+[["This", "is", "great!"],
+ ["Awesome!"]]
+```
+
+#### 3. Offset variants {#3-offset-variants}
+
+The tokenizers should include a companion op that returns byte offsets for each
+token into the originating string. We use *_with_offsets* function variants (as
+opposed to an additional *offsets=True* parameter) with the *start_offsets* &
+*limit_offsets* naming to match the style employed by
+`tf.strings.unicode_decode_with_offsets` and
+`tf.strings.unicode_split_with_offsets`. This also has the benefit of separating
+out the heavier alignment producing ops' return signatures in the API by name.
+
+#### 4. Vocabulary Lookups - tf.Hub {#4-vocabulary-lookups-tf-hub}
+
+For ops which require a vocabulary for looking up tokens, they should receive a
+lookup object, as opposed to a flat file containing the vocabulary.
+This can provide added functionality like out-of-vocab bucketing,
+allowing for varying lookup implementations, and better resource handling that a
+simple vocabulary file path cannot duplicate.
+
+For these ops requiring external assets, if there is a desire to include a
+standard version of the op along which includes the required asset files,
+these should be published as a [TF.Hub](https://www.tensorflow.org/hub) module.
+These modules can include popular versions of assets to be used for general use
+cases. The published TF.Hub module should continue to follow the other
+conventions outlined above.
+
+### New Tokenize APIs {#new-tokenize-apis}
+
+Following the above conventions, tokenization ops tend to follow a pattern.
+This is intentional, as it gives familiarity across differing implementations,
+and again allows for convenient exchanging of implementations.
+This is not a base class, but a general pattern that is followed throughout
+APIs of tokenization implementations following the above conventions.
+
+```python
+def xyz_tokenize(input, **kwargs):
```

Comment:
Why a function instead of an object?

I'd expect tokenizers to implement an interface like

```
class Tokenizer:
  def tokenize(input, encoding=None):

  def tokenize_with_offsets(input, encoding=None):
```

where any additional kwargs and tokenizer-specific configuration options are passed to the constructor, so it's easy to pass these objects around without all call sites having to also know what options to pass to tokenizers.

Reply:
Yeah; a valid question. Two main reasons:
- Many tokenizers past what are proposed here require a trained model and are not fit to be surfaced as ops. The expectation is that these conventions would apply to more than just ops, but also TF Hub modules and TF Serving models as well.
- While a majority of tokenizers could fit with the interface and having a constructor for initialization, I did not want to limit possible future methods. For example, a multilingual tokenizer might allow for a language_hint as an argument for tokenization, or you may want to pass different lookup tables to a wordpiece tokenizer depending on the language.

Reply:
> I am just bothered that if you're using these tokenizers in eager mode you'll need to keep passing initialization arguments again and again on every iteration. You can always make a free function a method by using a wrapper empty class, but you cannot in general go the other way without wasting work.

I have the same concern even in graph mode, where one may want to tokenize string inputs in multiple ops using essentially the same tokenizer. If a tokenizer depends on expensive initialization or external file resources (e.g. a look up table; or a lexicon), we would end up initializing/loading resources/allocating memory O(# ops) times instead of O(1).

Reply:
I've been discussing this subject offline as well and will update the RFC to include an interface.